CoreScriptsRoot/Modules/VR/LaserPointer.lua

```diff
@@ -0,0 +1,933 @@
+		local plopCF = CFrame.new(parabHitPoint, parabHitPoint + parabHitNormal)
+
+		--Calculate the height of the ball from a sine wave raised to a configurable exponent
+		if self.teleportValid then
+			local ballWave = applyExpCurve(math.sin((now * 2 * math.pi) / TELEPORT.BALL_WAVE_PERIOD), TELEPORT.BALL_WAVE_EXP)
+			ballHeight = TELEPORT.BALL_WAVE_START + (ballWave * TELEPORT.BALL_WAVE_AMPLITUDE)
+		else
+			ballHeight = 0--TELEPORT.BALL_WAVE_AMPLITUDE
+		end
+
+		self.plopPart.CFrame = plopCF
+		self.plopBall.CFrame = plopCF * CFrame.new(0, 0, -ballHeight)
+
+		--Handle the pulse animation
+		--We're basically scheduling it to begin every TELEPORT.PULSE_PERIOD seconds,
+		--and the animation runs for TELEPORT.PULSE_DURATION seconds. TELEPORT.PULSE_EXP
+		--affects the growth rate of the pulse size; ^2 is a good look, starts slow and accelerates.
+		local timeSincePulseStart = now % TELEPORT.PULSE_PERIOD
+		if timeSincePulseStart > 0 then
+			local pulseSize = timeSincePulseStart / TELEPORT.PULSE_DURATION
+			if pulseSize < 1 then
+				self.plopAdornPulse.Visible = true
+				self.plopAdornPulse.Size = identityVector2 * (TELEPORT.PULSE_SIZE_0 + applyExpCurve(pulseSize, TELEPORT.PULSE_EXP) * (TELEPORT.PULSE_SIZE_1 - TELEPORT.PULSE_SIZE_0))
+				self.plopAdornPulse.Transparency = 0.5 + (pulseSize * 0.5)
+			else
+				self.plopAdornPulse.Visible = false
+				self.plopAdornPulse.Size = zeroVector2
+				self.pulseStartTime = tick() + TELEPORT.PULSE_PERIOD
+			end
+		end
+	end
+
+
+	function LaserPointer:updateTeleportHoldState(dt)
+		--Increase launch velocity if teleport button is held
+		if self.teleportButtonDown and self.teleportMode then
+			self.teleportRangeT = math.min(1, self.teleportRangeT + dt)
+		else
+			self.teleportRangeT = 0
+		end
+	end
+
+	function LaserPointer:updateTeleportMode(hitPoint, hitNormal, hitPart)
+		self.teleportPoint = hitPoint
+		self.teleportNormal = hitNormal
+		self.teleportPart = hitPart
+
+		self:updateTeleportPlop(hitPoint, hitNormal)
+		self:setTeleportValidState(self:canTeleportTo(workspace.CurrentCamera.CFrame.p, self.teleportPart, self.teleportPoint, self.teleportNormal))
+	end
+end
+
+do --Event callbacks/update loop
+	function LaserPointer:update(dt)
+		if not self.enabled then
+			return
+		end
+		local ignore = { game.Players.LocalPlayer.Character, self.originPart, self.plopPart, self.plopBall }
+		local cameraSpace = workspace.CurrentCamera.CFrame
+		local thickness0, thickness1 = LASER.ARC_THICKNESS, TELEPORT.ARC_THICKNESS
+		local gravity0, gravity1 = LASER.G, TELEPORT.G
+		
+		self.parabola.Thickness = self.transitionAlpha * (thickness1-thickness0) + thickness0
+
+		if self:isHeadMounted() then
+			self.parabola.Thickness = self.parabola.Thickness * HEAD_MOUNT_THICKNESS_MULTIPLIER
+
+			--cast ray from center of camera, then render laser going from offset point to hit point
+			local originCFrame = cameraSpace * UserInputService:GetUserCFrame(Enum.UserCFrame.Head)
```

Comment:
Yeah, that's the plan. Once it ships for all platforms I'll come back and change all of them over. The behavior will remain the same.